Add Descript.com embeds support

alshakero opened the pull request:
Similar to #21101. Which was thoroughly reviewed.

#### Changes proposed in this Pull Request:

This PR adds support for Descript.com embedding.

1. It adds a Descript.com embedding block, you can test it by adding a "Descript" block and use [this URL](https://share.descript.com/view/gXEWtXaGTyK) to test.
2. Descript normal [URLs](https://share.descript.com/view/gXEWtXaGTyK) using the embed block.
3. Descript shortcodes like `[descript id="gXEWtXaGTyK" width="640" height="480"]`.
4. Descript embed code, it transforms it into a safe shortcode.
```html
<iframe src="https://share.descript.com/embed/jUxUmel6GyN" width="640" height="360" frameborder="0" allowfullscreen></iframe>
```

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
1. Apply D72574-code diff on your sandbox.
2. Sandbox a site. Create a new post.
3. Using the Descript block, embed https://share.descript.com/view/gXEWtXaGTyK.
3. Using the embed block, embed https://share.descript.com/view/gXEWtXaGTyK.
4. Using the shortcode block, please test `[descript id="gXEWtXaGTyK" width="640" height="480"]`.
5. Using custom HTML block, use the embed code they provide:
```html
<iframe src="https://share.descript.com/embed/jUxUmel6GyN" width="640" height="360" frameborder="0" allowfullscreen></iframe>
```

jeherve commented on projects/plugins/jetpack/modules/shortcodes/descript.php:
```diff
@@ -0,0 +1,134 @@
+	}
+
+	$params = array(
+		'id'     => $id,
+		'width'  => $width,
+		'height' => $height,
```
I think it would be nice if we could cast those as int / escape if possible.

jeherve commented on projects/plugins/jetpack/modules/shortcodes/descript.php:
```diff
@@ -0,0 +1,134 @@
+	// Allow script to be filtered to short code (so direct copy+paste can be done).
+	add_filter( 'pre_kses', 'jetpack_shortcodereverse_descript' );
+
+	// Descript rejects any request with the dnt query param, this removes the dnt from the oembed request.
+	add_filter( 'oembed_fetch_url', 'jetpack_descript_remove_dnt' );
```
This is a bit annoying. I wonder if we should opt to not embed, but only display a link when dnt is enabled? This way we would honor dnt at all times, and not just ignore it. I think someone who enabled dnt may prefer that over being tracked without knowing about it?

pdesantis replied:
Hello, I'm on the Descript engineering team. Good catch! We'll add support for the `dnt` parameter

alshakero replied:
Hey @pdesantis! That's awesome. I've sent you a support email.

pdesantis replied:
@alshakero just following up, the `dnt` parameter is now supported

jeherve commented on projects/plugins/jetpack/extensions/shared/icons.js:
```diff
@@ -275,6 +275,15 @@ export const SmartFrameIcon = {
+	src: (
+		<SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 43 43" xmlSpace="preserve">
+			<Path transform="translate(-25)" fillRule="evenodd" d="M37 31.908c0 1.376.893 2.268 2.27 2.268H45.9c3.956 0 7.192-1.337 9.392-3.68H37v1.412zM45.901 9H39.27C37.893 9 37 9.893 37 11.269v1.411h18.294C53.093 10.338 49.857 9 45.9 9zm5.787 16.176c0 1.113.726 1.835 1.845 1.835h3.973c.45-1.117.753-2.345.894-3.671h-4.867c-1.119 0-1.845.722-1.845 1.836zm-5.507-7.166c0 1.113.725 1.836 1.844 1.836H58.4a13 ...
+		</SVG>  
```
We'll need to remove that trailing space so we can merge this.

jeherve commented on projects/plugins/jetpack/extensions/extended-blocks/core-embed/descript.js:
```diff
@@ -0,0 +1,25 @@
+ * Internal dependencies
+ */
+import { DescriptIcon } from '../../shared/icons';
+/*
+ * New `core/embed` block variation.
+ */
+
```
Maybe it would be better like this?

```suggestion

/*
 * New `core/embed` block variation.
 */
```

jeherve reviewed (changes requested):
TIL about Descript! It seems nice. I left a bit of feedback below.

alshakero commented:
Thanks for the quick review! I addressed all feedback. 

To test the `dnt` change, please go to chrome://settings/cookies and try with "Do Not Track" on and off to see that when it's on, the shortcode is rendered as a link, when off, it's rendered as an iframe.

jeherve commented on projects/plugins/jetpack/modules/shortcodes/descript.php:
```diff
@@ -87,6 +87,20 @@ function jetpack_shortcodereverse_descript( $content ) {
+ *
+ * @return bool
+ */
+function jetpack_descript_is_dnt_enabled() {
```
We have something similar already:

https://github.com/Automattic/jetpack/blob/e50c798ac77031439d8cb6a4a632d29f70dfbe53/projects/plugins/jetpack/modules/stats.php#L95-L100

What do you think about extracting this to a shared location, maybe `projects/plugins/jetpack/functions.global.php` for now?

jeherve reviewed (commented):
Thanks for the quick turn-around. I have one more suggestion, since you're working on DNT?

alshakero commented:
Hi @pdesantis! Would it be possible to quickly whitelist the `dnt` flag rather than implementing actual `dnt` support as a first step? This will unlock this PR without any dirty hacks and removes the time pressure.

I've put this on hold considering the above.

jeherve commented on projects/plugins/jetpack/modules/shortcodes/descript.php:
```diff
@@ -0,0 +1,149 @@
+		esc_attr( $id )
+	);
+
+	// jetpack_is_dnt_enabled runs this filter before checking the headers.
+	add_filter( 'jetpack_honor_dnt_header_for_stats', '__return_true' );
```
I don't know that I would enable things here; I would let site owners make that choice. 

We can, if we'd like, enable it on WordPress.com Simple though.

yansern reviewed (dismissed):
Tested on both WPCOM and standalone WP site.
- Descript block works.
- Embed block with descript URL works.
- Shortcode block works (after enabling shortcodes through **Jetpack > Settings > Writing > Compose using Shortcodes**)
- Using `<iframe>` on Custom HTML block will not work on WPCOM (likely due to kses restriction, and that is OK). This works fine for standalone WP sites.

jeherve reviewed (changes requested):
This works well in my tests. 🚢

jeherve reviewed (dismissed):
This works well in my tests. 🚢
